### PR TITLE
Integrate dewolf-logic

### LINKED ISSUE #10: Integrate dewolf-logic
# Problem Statement 
We implemented our own logic engine (https://github.com/fkie-cad/dewolf-logic), which we want to integrate into dewolf. 

# Solution
Implement the logic interface and make sure that all needed simplifications are implemented in dewolf-logic.

### PR DESCRIPTION
**Pull request by ebehner**
closes #10

### CODE REVIEW

**Review by 0x6e62 — APPROVED** [review]

File: decompiler/structures/ast/ast_nodes.py [hunk 1]
```diff
@@ -163,10 +164,12 @@ def clean(self) -> None:
         """Makes clean ups, depending on the node. This helps to standardize the AST."""
         pass
 
-    def simplify_reaching_condition(self, z3_condition_of: Dict[LogicCondition, PseudoLogicCondition]):
+    def simplify_reaching_condition(self, condition_handler: Optional[ConditionHandler]):
         """Simplify the reaching condition. If it is false we remove the subtree of this node."""
+        if condition_handler is None:
+            return
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/ast/ast_nodes.py:170]: I removed it, I was trying to have it more general to be able to use it on the ast and asforest, but you are right, that makes no sense here

File: decompiler/structures/logic/custom_logic.py [hunk 2]
```diff
@@ -1,10 +1,534 @@
...
+    @classmethod
+    def generate_new_context(cls) -> World:
+        """Generate a context for z3-conditions."""
+        return World()
+
+    @property
+    def _condition(self) -> WorldObject:
+        if term := self.context.get_definition(self._variable):
+            return term
+        return self._variable
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:42]: I know, I have a `._condition` attribute for z3_logic. Here, with the custom logic, it is not an attribute, but for consistency I wanted it to be a property. That is why I chose to make it private. I can change it, not sure what is more "natural here"

File: decompiler/structures/logic/custom_logic.py [hunk 3]
```diff
@@ -1,10 +1,534 @@
...
+
+    @property
+    def is_symbol(self) -> bool:
+        """Check whether the object is a symbol."""
+        return self._is_symbol(self._condition)  # TODO
+
+    @property
+    def is_literal(self) -> bool:
+        """Check whether the object is a literal, i.e., a symbol or a negated symbol"""
+        return self._is_literal(self._condition)  # TODO
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:162]: I only forgot to remove them. Sometimes I mark stuff while working to easier find it. There are now removed

File: decompiler/structures/logic/custom_logic.py [hunk 4]
```diff
@@ -1,10 +1,534 @@
...
+            self.context.free_world_condition(self._variable)
+            self._variable.simplify()
+        else:
+            new_var = self.context.variable(f"Simplify", 1)
+            self.context.define(new_var, self._condition)
+            self.context.free_world_condition(new_var)
+            new_var.simplify()
+            self._variable = self.context.new_variable(1, tmp=True)
+            self.context.substitute(new_var, self._variable)
+        return self
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:221]: No, it does not raise any issues. The 'substitute' function removes the node 'new_variable'. Thus, it is only temporary in the graph and we can add it again for the next simplification.

File: decompiler/structures/logic/custom_logic.py [hunk 5]
```diff
@@ -1,10 +1,534 @@
...
+
+        subexpressions: List[LOGICCLASS] = [condition] if numb_of_arg_cond == 1 else condition_operands
+        for sub_expr_1, sub_expr_2 in product(subexpressions, self.operands):
+            if sub_expr_1.is_equivalent_to(sub_expr_2):
+                relations = self.context.get_relation(self._condition, sub_expr_2._condition)
+                for relation in relations:
+                    self.context.remove_operand(self._condition, relation.sink)
+        to_remove = [cond._variable for cond in condition_operands + operands if cond._variable != cond._condition]
+        self.context.cleanup(to_remove)
+        return self
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:276]: I split the function a bit. It is still quite long but more separation leads to functions with many parameters. Hope this is enpugh.

File: decompiler/structures/logic/custom_logic.py [hunk 6]
```diff
@@ -1,10 +1,534 @@
...
+
+    def remove_redundancy(self, condition_handler: ConditionHandler) -> LOGICCLASS:
+        """
+        More advanced simplification of conditions.
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:280]: I am not sure how to write this in imperative, so I changed it to a more specific Docstring.

File: decompiler/structures/logic/custom_logic.py [hunk 7]
```diff
@@ -1,10 +1,534 @@
...
+            else:
+                new_operands = list()
+                for op in operand.operands:
+                    if op in compared_expressions:
+                        new_operands.append(compared_expressions[op])
+                    else:
+                        assert isinstance(op, Constant), f"The operand must be a Constant"
+                        new_operands.append(pseudo.Constant(op.signed, pseudo.Integer(op.size, signed=True)))
+                condition_symbol = condition_handler.add_condition(Condition(self.OPERAND_MAPPING[operand.SYMBOL], new_operands))
+                self.context.replace(operand, condition_symbol.symbol._condition)
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:323]: Extracted the part that computes the symbol condition from the real condition to a separate function (lines 295 to323).

File: decompiler/structures/logic/custom_logic.py [hunk 8]
```diff
@@ -1,10 +1,534 @@
...
+
+        self.context.replace(self._condition, real_condition._condition)
+        self.context.cleanup()
+        return self
+
+    def _replace_symbols_by_real_conditions(self, condition_handler: ConditionHandler):
+        """
+        Return the real condition where the symbols are replaced by the conditions of the condition handler and all
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:331]: updated

File: decompiler/structures/logic/custom_logic.py [hunk 9]
```diff
@@ -1,10 +1,534 @@
...
+        compared_expressions: Dict[Variable, pseudo.Expression] = dict()
+        for symbol in self.get_symbols():
+            pseudo_condition: Condition = condition_handler.get_condition_of(symbol)
+            for operand in pseudo_condition.operands:
+                if not isinstance(operand, pseudo.Constant):
+                    compared_expressions[self.context.variable(self._variable_name_for(operand))] = operand
+            self._replace_symbol(symbol, condition_handler, condition_nodes)
+        return copied_condition, compared_expressions
+
+    def _replace_symbol(self, symbol: CustomLogicCondition, condition_handler: ConditionHandler, condition_nodes: Set[WorldObject]):
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:345]: done

File: decompiler/structures/logic/custom_logic.py [hunk 10]
```diff
@@ -1,10 +1,534 @@
...
+        """Get literals on World-level"""
+        if self._is_literal(condition):
+            yield condition
+        elif isinstance(condition, (BitwiseOr, BitwiseAnd, BitwiseNegate)):
+            for child in condition.operands:
+                yield from self._get_literals(child)
+        else:
+            assert isinstance(condition, Constant) and condition.size == 1, f"The condition {condition} does not consist of literals."
+
+    def _rich_string_representation(self, condition: WorldObject, condition_map: Dict[Variable, pseudo.Condition]):
```
[the remark this answers is not in the document]
Comment (ebehner) [reply decompiler/structures/logic/custom_logic.py:404]: done

**Pull request merged**